# 承诺

> 原文:[https://www.learn-js.org/en/Promises](https://www.learn-js.org/en/Promises)

* * *

承诺是 JavaScript 异步编程的基础，掌握它非常重要。

### 什么是异步编程？

异步编程，简称为 async 编程，是一种使代码的不同部分能够在不断变化的时间运行而不是立即运行的编程方法。

当我们想从某个远程服务器获取信息，并编写代码对服务器返回的信息进行处理时，这是最需要的:

```
function getServerStatus() {
    const result = fetch("/server/status");

    // THIS WILL NOT WORK!
    console.log("The status from the server is: ", result.ok);
} 
```

在 Python 等许多编程语言中，这种方法是可行的，因为函数在默认情况下是同步函数。

在 JavaScript 中，大多数需要等待函数做某事的 API 在默认情况下是**异步的**，这意味着这段代码不会做我们认为它会做的事情，因为`fetch`函数是异步的，因此会返回一些不完全是结果的东西，但是*最终*会是结果吗？从`fetch`函数返回的这个“东西”在 JavaScript 中被称为**承诺**。

为了让上面的代码工作，我们需要按照下面的方式编写函数:

```
function getServerStatus() {
    const result = fetch("/server/status");

    result.then(function(status) {
        console.log("The status from the server is: ", status.ok);
    });
} 
```

注意，我们在这里使用了`then`函数，它是`Promise`的方法之一。

### 许诺对象

Promise 是一个本地 JavaScript 对象，它有两个特征:1 .它接收一个函数参数。这个函数需要有两个参数，一个`resolve`函数和一个`reject`函数。写在 promise 里面的代码需要使用这两个函数中的一个。2.可以使用`then`方法(和其他类似的方法)或`await`语句来等待。(async/await 语句有单独的教程)。

异步函数是由一个函数定义的，该函数不是返回它应该返回的值，而是返回一个`Promise`对象，该对象将最终解析并给用户答案。

例如，假设我们想要计算两个数的和，但是通过编写一个函数返回一个`Promise`而不是值。

```
function sumAsync(x, y) {
    const p = new Promise((resolve, reject) => {
        // this resolves the promise we just created with the output of x+y
        resolve(x + y);                        
    });

    // This returns the promise, not the value
    return p;
}

// let's use the function now
sumAsync(5, 7).then((result) => {
    console.log("The result of the addition is:", result);
}); 
```

这在什么时候会非常有用？当需要间接进行计算时，例如在等待一段时间后，或者使用`fetch`命令从服务器检索信息时。

让我们修改示例，仅在半秒钟后求解:

```
function sumAsync(x, y) {
    console.log("1\. sumAsync is executed");
    const p = new Promise((resolve, reject) => {
        // run this in 500ms from now
        setTimeout(() => {
            console.log("4\. Resolving sumAsync's Promise with the result after 500ms");
            resolve(x + y);
        }, 500);

        // we don't need to return anything
        console.log("2\. sumAsync Promise is initialized");            
    });
    console.log("3\. sumAsync has returned the Promise");
    return p;
}

// let's use the function now
sumAsync(5, 7).then((result) => {
    console.log("5\. The result of the addition is:", result);
}); 
```

### 拒绝承诺

在同步流中，如果我们想告诉用户发生了错误，这样他就可以捕捉到异常，我们使用`throw`参数抛出异常。当使用承诺时，我们需要改为触发`reject`函数。

假设我们想写一个相同的函数，但是如果一个值是负的，我们就拒绝这个函数:

```
function sumAsync(x, y) {
    return new Promise((resolve, reject) => {
        // run this in 500ms from now
        setTimeout(() => {
            if (x < 0 || y < 0) {
                reject("Negative values received");
            } else {
                resolve(x + y);
            }
        }, 500);

        // we don't need to return anything
    });
}

sumAsync(-5, 7).then((result) => {
    console.log("The result of the addition is:", result);
}).catch((error) => {
    console.log("Error received:", error);
}); 
```

## 锻炼

写一个函数，接收一个字符串，并返回一个承诺。

promise 应该用字符串的大写版本进行解析，但是如果字符串为空，则应该拒绝。